### PR TITLE
Support spmd mode for e2e alignment tasks.

### PR DESCRIPTION
**Pull request by charles9304**

### CODE REVIEW

File: chatlearn/models/base_module.py [hunk 1]
```diff
@@ -485,7 +502,8 @@ def setup_collective_group(self, rank, world_size, backend, group_name):
         """
         self._group_names.append(group_name)
         self._world_size = world_size
-        col.init_collective_group(
+        # breakpoint()
```
Comment (haolin-nju) [inline chatlearn/models/base_module.py:505]: debug code can be deleted

File: chatlearn/models/base_module.py [hunk 2]
```diff
@@ -767,18 +790,39 @@ def broadcast_parameter(self, rank, src_rank, group_name, pipe_stage=0):
         :meta private:
         """
         tensors = []
+        # breakpoint()
```
Comment (haolin-nju) [inline chatlearn/models/base_module.py:793]: same here

File: chatlearn/models/base_module.py [hunk 3]
```diff
@@ -767,18 +790,39 @@ def broadcast_parameter(self, rank, src_rank, group_name, pipe_stage=0):
...
+        print(f"debug from {src_rank} to {rank} self._parameters_to_sync[{pipe_stage}]: {len(self._parameters_to_sync[pipe_stage])} {self._parameters_to_sync[pipe_stage][0][0]}", flush=True)
```
Comment (haolin-nju) [inline chatlearn/models/base_module.py:794]: a debug level msg should not be printed by `print`

File: examples/megatron/configs/llama2/vllm_param_sync.yaml [hunk 4]
```diff
@@ -49,3 +50,4 @@ runtime:
   exp_name: ${exp_name:chatlearn}
   debug: ${debug:False}
   validate_param_sync: ${validate_param_sync:False}
+  param_sync_comm_type: ${param_sync_comm_type:broadcast}
```
Comment (haolin-nju) [inline examples/megatron/configs/llama2/vllm_param_sync.yaml:53]: what's this for?

File: chatlearn/synchronizer/parameter_sync.py [hunk 5]
```diff
@@ -1045,33 +1058,47 @@ def _multi_thread_sync_for_tp_num_mapping_eq_1(
...
+        #                 futures.append(executor.submit(
+        #                     self.sync_send_recv, send_actor, recv_actor, requires_grad, filter_fn=filter_fn, param_group=param_group
+        #                 ))
+        #     for _future in concurrent.futures.as_completed(futures):
+        #         try:
+        #             _future.result()œ
+        #         except Exception as e:
+        #             raise RuntimeError(f"Parameter sync thread generated an exception: {e}") # pylint: disable=raise-missing-from
+        #     concurrent.futures.wait(futures)
+        if True:
```
Comment (haolin-nju) [inline chatlearn/synchronizer/parameter_sync.py:1091]: formal code should be uncommented?

**Review by haolin-nju — COMMENTED** [review]
See comments. Some debug msgs and code should be revised.

File: chatlearn/models/base_module.py [hunk 6]
```diff
@@ -39,6 +39,22 @@
 from chatlearn.launcher import dlc_utils
 
 
+from pynvml import nvmlInit, nvmlDeviceGetCount, nvmlDeviceGetHandleByIndex, nvmlDeviceGetComputeRunningProcesses, nvmlShutdown
+
+def get_gpu_usage():
```
Comment (adoda) [inline chatlearn/models/base_module.py:44]: Only used for debug info

File: examples/megatron/configs/llama2/vllm_rlhf.yaml [hunk 7]
```diff
@@ -64,7 +64,8 @@ models:
 
 runtime:
   colocation:
-    - policy,ppo_policy,reward,reference,value,ppo_value
+    - policy
```
Comment (adoda) [inline examples/megatron/configs/llama2/vllm_rlhf.yaml:67]: don't change the default config.

File: examples/megatron/models/vllm_policy_inference.py [hunk 8]
```diff
@@ -138,21 +138,22 @@ def decode_internal(self, batched_outputs):
...
         loss_mask = get_loss_mask(all_tokens, self.tokenizer.tokenizer.eos_token_id, prompt_sizes)
         loss_mask = loss_mask.to("cpu")
+        print(f"str_outputs: {len(str_outputs)} {str_outputs}")
         return {"all_tokens": all_tokens, "str_outputs": str_outputs, "str_prompts": str_prompts,
             "no_padded_query_ids": no_padded_query_ids, "logprobs": logprobs,
             "loss_mask": loss_mask}
 
-class VLLMPolicyInferenceAsync(VLLMPolicyInference):
-    """VLLMPolicyInferenceAsync is the model for VLLMModuleV2, which uses async generate API"""
+class VLLMPolicyInferenceV2(VLLMPolicyInference):
```
Comment (adoda) [inline examples/megatron/models/vllm_policy_inference.py:146]: It's the same to VLLMPolicyInference, we can delete it

File: chatlearn/models/vllm_module_v2.py [hunk 9]
```diff
@@ -146,17 +181,17 @@ def _convert_v1_inputs(self, prompts, prompt_token_ids):
...
         input_ids_key = self.model_args.get("vllm_input_ids_key", "input_ids")
 
         prompts = query[prompt_key]
         prompts_token_ids = query[input_ids_key]
         seq_len = self.model_args.get("seq_length")
-        final_outputs = []
         parsed_prompts = []
         sampling_params = []
         for i, prompt in enumerate(prompts):
+            request_id = i
```
Comment (adoda) [inline chatlearn/models/vllm_module_v2.py:194]: unused request_id

File: chatlearn/models/vllm_module_v2.py [hunk 10]
```diff
@@ -180,37 +215,16 @@ async def generate_vllm(self, query, is_eval):
             sampling_params,
             use_tqdm=True,
         )
-        final_outputs = sorted(outputs, key=lambda x: int(x.request_id))
```
Comment (adoda) [inline chatlearn/models/vllm_module_v2.py:183]: why delete sorted

File: examples/megatron/models/vllm_policy_inference.py (hunk 8 above)
Comment (charles9304) [reply examples/megatron/models/vllm_policy_inference.py]: Done

File: examples/megatron/configs/llama2/vllm_rlhf.yaml (hunk 7 above)
Comment (charles9304) [reply examples/megatron/configs/llama2/vllm_rlhf.yaml]: done

File: chatlearn/synchronizer/parameter_sync.py (hunk 5 above)
Comment (charles9304) [reply chatlearn/synchronizer/parameter_sync.py]: Revert already.

File: examples/megatron/configs/llama2/vllm_param_sync.yaml (hunk 4 above)
Comment (charles9304) [reply examples/megatron/configs/llama2/vllm_param_sync.yaml]: Revert already.

File: chatlearn/models/base_module.py (hunk 3 above)
Comment (charles9304) [reply chatlearn/models/base_module.py]: Delete already.

File: chatlearn/models/base_module.py (hunk 2 above)
Comment (charles9304) [reply chatlearn/models/base_module.py]: Delete already.

File: chatlearn/models/base_module.py (hunk 1 above)
Comment (charles9304) [reply chatlearn/models/base_module.py]: Delete already.

File: chatlearn/models/vllm_module_v2.py (hunk 10 above)
Comment (charles9304) [reply chatlearn/models/vllm_module_v2.py:183]: LLM.generate API already sorts outputs.

File: chatlearn/models/vllm_module_v2.py (hunk 9 above)
Comment (charles9304) [reply chatlearn/models/vllm_module_v2.py]: Delete already.

**Review by adoda — APPROVED** [review]

**Review by haolin-nju — APPROVED** [review]
LGTM

**Pull request merged**